docs: update quickstart and references for embedded replica updates (#144)
tursodatabase/turso-docs: sdk/rust/reference.mdx

---
title: Reference
description: libSQL Rust Reference
---

> The libSQL Rust crate contains everything you need to work with Turso and asynchronous Rust.

Reviewer: ```suggestion
The libSQL Rust crate contains everything you need to work with Turso and works flawlessly with popular async runtimes like `tokio`.
```

## Installing

> Install the package in your project using the following command:

Reviewer: ```suggestion
Install the crate in your project using the following command:
```

```bash
cargo add libsql
```

## Initializing

> Make sure add crate to your project at the top of your file:

Reviewer: ```suggestion
Make sure add the crate to your project at the top of your file:
```

---
title: Reference
description: libSQL Rust Reference
---

The libSQL Rust crate contains everything you need to work with Turso and asynchronous Rust.

## Installing

Install the package in your project using the following command:

```bash
cargo add libsql
```

> 

Reviewer: ```suggestion

### Conditional compilation

The libsql rust client supports conditionally compiling certain features to reduce compile times depending on what features you would like to use. Currently, all features are enabled by default but it is possible to set `default-features = false` and to enable only the ones you would like. The available features are `remote` (http only client), `core` (local database only), `replication` (embedded replica features) and `encryption` (encryption at rest support, this is disabled by default). For more information on how to enable features you checkout the [cargo book](https://doc.rust-lang.org/cargo/reference/features.html#dependency-features).

```

Author: @LucioFranco Would this example be sufficient here on how to customise this?

```toml
[dependencies]
libsql = { version = "...", "default-features" = false, features = ["encryption"] }
```

```rust
use libsql::Builder;
use std::time::Duration;

let url = env::var("LIBSQL_URL").expect("LIBSQL_URL must be set");
let token = env::var("LIBSQL_AUTH_TOKEN").unwrap_or_default();

let mut db = Builder::new_remote_replica("local.db", &url, &token)
  .periodic_sync(Duration::from_secs(300)) // Sync every 5 minutes
  .build()
  .await.unwrap();

let conn = db.connect().unwrap();
```

### Read Your Own Writes

The `read_your_writes` function configures the database connection to ensure that writes are immediately visible to subsequent read operations initiated by the same connection. This is enabled by default and is particularly important in distributed systems to ensure consistency from the perspective of the writing process.

You can disable this behavior by passing `false` to the function:

```rust
use libsql::Builder;

let url = env::var("LIBSQL_URL").expect("LIBSQL_URL must be set");
let token = env::var("LIBSQL_AUTH_TOKEN").unwrap_or_default();

let mut db = Builder::new_remote_replica("local.db", &url, &token)
  .read_your_writes(false) // Disable reading your own writes
  .build()
  .await.unwrap();

let conn = db.connect().unwrap();
```

## Encryption

> To enable encryption on a SQLite file (`new_local` or `new_remote_replica`), pass the `encryption_config`:

Reviewer: make sure to mention this requires enabling the `encryption` feature which is disabled by default.

Author: Ah, @LucioFranco this is a great catch. I'm going to further tweak the examples to show how to enable the features with `features = ["encryption"]` etc.

```rust
use libsql::Builder;
use bytes::Bytes;

let url = env::var("LIBSQL_URL").expect("LIBSQL_URL must be set");
let token = env::var("LIBSQL_AUTH_TOKEN").unwrap_or_default();

let cipher = Cipher::YourChosenCipher;
let encryption_key_bytes = Bytes::from("your_secure_encryption_key_here");

let encryption_config = EncryptionConfig {
    cipher,
    encryption_key: encryption_key_bytes,
};

let mut db = Builder::new_remote_replica("local.db", &url, &token)
  .encryption_config(encryption_config) // Apply encryption configuration
  .build()
  .await.unwrap();

let conn = db.connect().unwrap();
```

<Info>

Encrypted databases appear as raw data and cannot be read as standard SQLite databases. You must use the libSQL client for any operations — [learn more](/libsql#encryption-at-rest).

</Info>

## Simple query

You can pass a string to `execute()` to invoke a SQL statement, as well as optional arguments:

<CodeGroup>

```rust Query

> conn.execute("SELECT * FROM users").await?;

Reviewer: ```suggestion
conn.execute("SELECT * FROM users", ()).await?;
```

```

```rust Arguments
conn.execute("SELECT * FROM users WHERE id = ?1", [1]).await?;
```

</CodeGroup>

## Prepared Statements

You can prepare a cached statement using `prepare()` and then execute it with `query()`:

```rust
let stmt = db_conn.prepare("SELECT * FROM users").await?;

> db_conn.query(&stmt, &[&1]).await?;

Reviewer: ```suggestion
db_conn.query(&stmt, [&1]).await?;
```

Reviewer: I think this should work, would be good to verify.

```

## Placeholders

libSQL supports the use of positional and named placeholders within SQL statements:

<CodeGroup>

```rust Positional
conn.execute("SELECT * FROM users WHERE id = ?1", params![1]).await?;
```

```rust Named
conn.execute("INSERT INTO users (name) VALUES (:name)", libsql::named_params! { ":name": "Iku" }).await?;
```

</CodeGroup>

## Deserialization

You can use the `de::from_row` function to deserialize a row into a struct:

```rust
use libsql::{de, Builder};

let mut stmt = conn
  .prepare("SELECT * FROM users WHERE id = ?1")
  .await
  .unwrap();
let row = stmt
  .query([1])
  .await
  .unwrap()
  .next()
  .await
  .unwrap()
  .unwrap();

#[derive(Debug, serde::Deserialize)]

> #[allow(dead_code)]

Reviewer: ```suggestion
```

struct User {
  name: String,
  age: i64,
  vision: f64,
  avatar: Vec<u8>,
}

let user = de::from_row::<User>(&row).unwrap();
```

## Batch Transactions

A batch consists of multiple SQL statements executed sequentially within an implicit transaction. The backend handles the transaction: success commits all changes, while any failure results in a full rollback with no modifications.

```rust
conn.execute_batch(r#"
  CREATE TABLE IF NOT EXISTS users (
    id INTEGER PRIMARY KEY,
    name TEXT NOT NULL
  );

  INSERT INTO users (name) VALUES ('Alice');
  INSERT INTO users (name) VALUES ('Bob');
"#).await?;
```

> ## Interactive Transactions

Reviewer: would be good to mention for embedded replica usage we also support a `BEGIN READ ONLY` transaction type that will not allow writes and will forward all the queries to the local db instead of the remote one. Any transaction that is a write transaction will all get fowarded to the primary remote db.

| ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `remote`      | Enables the HTTP-only client, allowing communication with a remote sqld server using pure Rust. Does not require compiling C code for SQLite. Suitable for projects that only need to interact with a remote database.                               |
| `core`        | Enables the local database only, incorporating the C SQLite3 code into the build. This is the foundation for local database operations but does not include additional features like replication or encryption.                                      |
| `replication` | Combines core with additional code required for replication, enabling the embedded replica features.                                                                                                                                                 |

> | `encryption`  | Enables encryption at rest support, adding the necessary code to compile encryption capabilities and expose functions for configuring it. **This is optional and not enabled by default**, catering to projects that require enhanced data security. |

Reviewer: I would add this this requires cmake installed on your system to work.
